src/neuroglancer/datasource/computed/backend.ts

```diff
@@ -0,0 +1,594 @@
+  private reject_?: (reason: Error) => void;
+
+  // True iff this chunk is actively computing.
+  private computing_ = false;
+
+  // True iff this chunk has been initialized for computation.
+  private initialized_ = false;
+
+  // A map from grid position string keys, as returned by gridPositionKey to
+  // vec3 grid positions. This is used as an indirection to avoid storing
+  // explicit references to VolumeChunks belonging to the origin source.
+  private originGridPositions_ = new Map<String, vec3>();
+
+  // The lower bound of the input patch.
+  private inputLower_?: vec3;
+
+  // Represents the input patch.
+  private inputBuffer_?: ArrayBuffer;
+
+  // Indicate to the ChunkManager that this is a computational Chunk.
+  isComputational = true;
+
+  // Our source is a ComputedVolumeChunkSource.
+  source: ComputedVolumeChunkSource;
+
+  /**
+   * Sets up computation parameters, computes overlapping origin chunks and
+   * initializes the input buffer. Returns a Promise that will resolve when
+   * computation completes, or reject if computation fails or is cancelled.
+   * @param computationParams computation parameters
+   * @param cancellationToken cancellation token
+   */
+  initializeComputation(
+      computationParams: ComputationParameters, cancellationToken: CancellationToken) {
+    if (!this.source) {
+      throw new Error('initializeComputation must be called after source is valid.');
+    }
+    if (!this.chunkDataSize) {
+      throw new Error('initializeComputation must be called after computeChunkBounds.');
+    }
+    this.computationParams_ = computationParams;
+    this.cancellationToken_ = cancellationToken;
+    this.computing_ = false;
+
+    this.cancellationToken_.add(() => {
+      if (this.reject_) {
+        this.fail_(CANCELED);
+      }
+    });
+    // Compute the input bounding box for this manager
+    // These computations happen without regard for edge effects, which are
+    // handled post-computation by cropping to this VolumeChunk's geometry.
+    const {inputSpec, outputSpec} = this.computationParams_;
+    const twos = [2.0, 2.0, 2.0];
+    const outBoxLower = vec3.multiply(vec3.create(), this.chunkGridPosition, outputSpec.size);
+    const outputCenter =
+        vec3.add(vec3.create(), outBoxLower, vec3.divide(vec3.create(), outputSpec.size, twos));
+    const scaleFactor = this.source.parameters.scaleFactor;
+    const inputCenter = vec3.divide(vec3.create(), outputCenter, scaleFactor);
+    const inputSize = inputSpec.size;
+    this.inputLower_ =
+        vec3.subtract(vec3.create(), inputCenter, vec3.divide(vec3.create(), inputSize, twos));
+    const bufferLength = prod(inputSize) * inputSpec.numChannels;
+    const originDataType = inputSpec.dataType;
+
+    // Signal that we're now taking up memory. This value will be overwritten
+    // post-computation by a call to decodeRawChunk.
+    this.systemMemoryBytes = bufferLength * DATA_TYPE_BYTES[originDataType];
```

Comment:
It is nice that this code is able to account for the memory used.  However, I think this will lead to systemMemoryCapacity accounting being incorrect, because the current code doesn't expect the chunk.systemMemoryBytes value to change while in a queue state where chunk.systemMemoryBytes is included in systemMemoryCapacity.  Previously, the systemMemoryBytes would only change while in DOWNLOADING state, which was not included in systemMemoryCapacity.  The same issue would also apply when calling decodeRawChunk later.  To fix this, we probably need a separate variable, like newPriorityTier, newPriority, etc., and then only actually update systemMemoryBytes when the queue state changes.  Then you could also set systemMemoryBytes in the chunk constructor rather than in initializeComputation since the transition to COMPUTING state happens before initializeComputaiton is called.

Reply:
The problem I see with this approach is that the chunk has basically no context in the constructor. I think I'd prefer to add a function call on the chunk object that returns an estimate of the memory it would occupy, perhaps even at a given state, replacing the systemMemoryBytes variable. Or, I could see implementing systemMemoryBytes as a getter in this one specific case.

Reply:
Good point, the constructor is too soon.  But you could override initializeVolumeChunk.  After you call super.initializeVolumeChunk, I think all relevant parameters should be present.